File: kakao.app/js/ked/view/editor/fileeditor.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, isStr: function (o) {return typeof o === 'string' || o instanceof String}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, isArr: function (o) {return Array.isArray(o)}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, copy: function (o) { return Array.isArray(o) ? o.slice() : typeof o == 'object' && o.constructor.name == 'Object' ? Object.assign({}, o) : typeof o == 'string' ? ''+o : o }, clamp: function (l,h,v) { var ll = Math.min(l,h), hh = Math.max(l,h); if (!_k_.isNum(v)) { v = ll }; if (v < ll) { v = ll }; if (v > hh) { v = hh }; if (!_k_.isNum(v)) { v = ll }; return v }, isNum: function (o) {return !isNaN(o) && !isNaN(parseFloat(o)) && (isFinite(o) || o === Infinity || o === -Infinity)}}

var fileeditor

import kxk from "../../../kxk.js"
let post = kxk.post
let kstr = kxk.kstr
let slash = kxk.slash

import nfs from "../../../kxk/nfs.js"

import belt from "../../edit/tool/belt.js"

import editor from "../../edit/editor.js"

import fileutil from "../../util/fileutil.js"

import indexer from "../../index/indexer.js"

import view from "../base/view.js"

import context from "../menu/context.js"

import mapscr from "./mapscr.js"


fileeditor = (function ()
{
    _k_.extend(fileeditor, editor)
    function fileeditor (screen, name)
    {
        var features

        this["onWheel"] = this["onWheel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["jumpToCounterpart"] = this["jumpToCounterpart"].bind(this)
        this["jumpToWord"] = this["jumpToWord"].bind(this)
        this["onContextChoice"] = this["onContextChoice"].bind(this)
        this["onContext"] = this["onContext"].bind(this)
        this["onGotoEof"] = this["onGotoEof"].bind(this)
        this["onGotoBof"] = this["onGotoBof"].bind(this)
        this["onGotoLine"] = this["onGotoLine"].bind(this)
        this["onGitDiff"] = this["onGitDiff"].bind(this)
        this["onGitCommit"] = this["onGitCommit"].bind(this)
        this["setCurrentFile"] = this["setCurrentFile"].bind(this)
        features = ['scroll','gutter','mapscr','complete','filepos','replex','brckts','unype','salter','vimple','uniko']
        fileeditor.__super__.constructor.call(this,screen,name,features)
        if (this.feats.mapscr)
        {
            this.mapscr = new mapscr(this)
            this.mapscr.show()
        }
        post.on('editor.highlight',this.state.highlightText)
        post.on('goto.line',this.onGotoLine)
        post.on('goto.bof',this.onGotoBof)
        post.on('goto.eof',this.onGotoEof)
        post.on('git.diff',this.onGitDiff)
        post.on('git.commit',this.onGitCommit)
    }

    fileeditor.prototype["setCurrentFile"] = function (currentFile)
    {
        this.currentFile = currentFile
    
        var _53_15_

        this.gutter.gitChanges = {}
        return (this.mapscr != null ? this.mapscr.reload() : undefined)
    }

    fileeditor.prototype["onGitCommit"] = function ()
    {
        return this.gutter.gitChanges = {}
    }

    fileeditor.prototype["onGitDiff"] = function (diff)
    {
        var currentFile

        currentFile = ked_session.get('editor▸file')
        if (diff.file === currentFile)
        {
            return this.gutter.onGitDiff(diff)
        }
    }

    fileeditor.prototype["onGotoLine"] = function (row, col, view)
    {
        var mc

        mc = this.state.mainCursor()
        col = (col != null ? col : mc[0])
        if (_k_.isStr(col))
        {
            switch (col)
            {
                case 'ind':
                    col = belt.numIndent(this.state.s.lines[row])
                    break
            }

        }
        if (!_k_.empty(view))
        {
            this.state.setView(view)
            return this.state.setCursors([[col,row]])
        }
        else
        {
            return this.state.setCursors([[col,row]],{adjust:'topBotDelta'})
        }
    }

    fileeditor.prototype["onGotoBof"] = function ()
    {
        return this.state.moveCursors('bof')
    }

    fileeditor.prototype["onGotoEof"] = function ()
    {
        return this.state.moveCursors('eof')
    }

    fileeditor.prototype["onContext"] = function (event)
    {
        var word

        word = this.state.textOfSelectionOrWordAtCursor()
        if (!_k_.empty(word))
        {
            word = ` '${word}'`
        }
        return context.show(event.cell,this.onContextChoice,[`search${word}`,`find${word}`,'status'])
    }

    fileeditor.prototype["onContextChoice"] = function (choice)
    {
        if (choice.startsWith('search'))
        {
            return post.emit('searcher.show',kstr.trim(choice.slice(6, -1)," '"))
        }
        else if (choice.startsWith('find'))
        {
            return post.emit('finder.show',kstr.trim(choice.slice(4, -1)," '"))
        }
        else
        {
            switch (choice)
            {
                case 'status':
                    return post.emit('differ.status')

            }

        }
    }

    fileeditor.prototype["jumpToWord"] = function (word)
    {
        var clss, currentFile, fun, func, idx

        if (clss = indexer.singleton.classes[word])
        {
            post.emit('file.open',clss.file,clss.line - 1)
            return true
        }
        else if (func = indexer.singleton.funcs[word])
        {
            if (_k_.isArr(func))
            {
                currentFile = ked_session.get('editor▸file')
                var list = _k_.list(func)
                for (idx = 0; idx < list.length; idx++)
                {
                    fun = list[idx]
                    if (fun.file === currentFile)
                    {
                        func = func[(idx + 1) % func.length]
                        break
                    }
                }
                if (_k_.isArr(func))
                {
                    func = func[0]
                }
            }
            if (!_k_.empty(func.file))
            {
                post.emit('file.open',func.file,func.line - 1,'ind')
                return true
            }
        }
        else
        {
            console.log(`fileeditor.jumpToWord(${word}) nothing found to jump to`)
        }
        return false
    }

    fileeditor.prototype["jumpToCounterpart"] = async function ()
    {
        var counter, currentFile, currext, ext, file, _173_50_, _179_50_, _188_50_

        currentFile = ked_session.get('editor▸file')
        currext = slash.ext(currentFile)
        var list = ((_173_50_=fileutil.counterparts[currext]) != null ? _173_50_ : [])
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            ext = list[_a_]
            if (await nfs.fileExists(slash.swapExt(currentFile,ext)))
            {
                post.emit('file.open',slash.swapExt(currentFile,ext))
                return
            }
        }
        var list1 = ((_179_50_=fileutil.counterparts[currext]) != null ? _179_50_ : [])
        for (var _b_ = 0; _b_ < list1.length; _b_++)
        {
            ext = list1[_b_]
            counter = slash.swapExt(currentFile,ext)
            file = fileutil.swapLastDir(counter,currext,ext)
            if (await nfs.fileExists(file))
            {
                post.emit('file.open',file)
                return
            }
        }
        var list2 = ((_188_50_=fileutil.counterparts[currext]) != null ? _188_50_ : [])
        for (var _c_ = 0; _c_ < list2.length; _c_++)
        {
            ext = list2[_c_]
            counter = slash.swapExt(currentFile,ext)
            if (_k_.in(currext,['noon']))
            {
                file = fileutil.swapLastDir(counter,'kode','js')
                if (await nfs.fileExists(file))
                {
                    post.emit('file.open',file)
                    return
                }
            }
            if (_k_.in(currext,['json']))
            {
                file = fileutil.swapLastDir(counter,'js','kode')
                if (await nfs.fileExists(file))
                {
                    post.emit('file.open',file)
                    return
                }
            }
        }
        console.log('cant find counterpart',currentFile)
    }

    fileeditor.prototype["onMouse"] = function (event)
    {
        var col, ret, row, start, word, x, y, _247_41_, _305_31_

        ret = fileeditor.__super__.onMouse.call(this,event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        var _a_ = this.eventPos(event); col = _a_[0]; row = _a_[1]

        switch (event.type)
        {
            case 'press':
                if (event.count > 1)
                {
                    if (!event.shift && event.button === 'left')
                    {
                        this.state.deselect()
                    }
                    x = col + this.state.s.view[0]
                    y = row + this.state.s.view[1]
                    this.state.clearHighlights()
                    if (event.count === 2)
                    {
                        if (event.alt)
                        {
                            this.state.selectChunk(x,y)
                        }
                        else
                        {
                            this.state.selectWord(x,y)
                        }
                    }
                    else
                    {
                        this.state.selectLine(y)
                    }
                    this.state.highlightSelection()
                    this.dragStart = _k_.copy(this.state.s.selections[0])
                    return {redraw:true}
                }
                else if (this.hover || (this.gutter != null ? this.gutter.cells.isInsideEvent(event) : undefined))
                {
                    x = col + this.state.s.view[0]
                    y = row + this.state.s.view[1]
                    if (event.cmd || event.ctrl)
                    {
                        if (word = belt.wordAtPos(this.state.s.lines,[x,y]))
                        {
                            console.log(`jumpToWord ${word}`,event)
                            if (this.jumpToWord(word))
                            {
                                return
                            }
                        }
                    }
                    this.dragStart = [x,y,x]
                    if (!event.shift && event.button === 'left')
                    {
                        this.state.deselect()
                    }
                    if (!event.alt)
                    {
                        this.state.clearCursors()
                    }
                    if (event.alt)
                    {
                        this.state.addCursor(x,y)
                    }
                    else
                    {
                        if (event.shift && this.state.s.cursors.length === 1)
                        {
                            this.state.setMainCursorAndSelect(x,y)
                        }
                        else
                        {
                            this.state.setMainCursor(x,y)
                        }
                    }
                    this.grabFocus()
                    return {redraw:true}
                }
                break
            case 'drag':
                if (this.dragStart)
                {
                    x = col + this.state.s.view[0]
                    y = row + this.state.s.view[1]
                    start = [this.dragStart[0],this.dragStart[1]]
                    if (y < this.dragStart[1])
                    {
                        start = [this.dragStart[2],this.dragStart[1]]
                    }
                    if (event.shift)
                    {
                        this.state.addRangeToSelectionWithMainCursorAtEnd(belt.rangeFromStartToEnd(start,[x,y]))
                    }
                    else
                    {
                        this.state.select(start,[x,y])
                    }
                    return {redraw:true}
                }
                break
            case 'release':
                delete this.dragStart
                break
            case 'move':
                if (this.hover)
                {
                    if (!this.hasFocus() && _k_.empty(view.currentPopup))
                    {
                        this.grabFocus()
                    }
                    post.emit('pointer','text')
                }
                else if ((this.gutter != null ? this.gutter.cells.isInsideEvent(event) : undefined))
                {
                    post.emit('pointer','vertical-text')
                }
                break
        }

        return this.hover
    }

    fileeditor.prototype["onWheel"] = function (event)
    {
        var start, steps, x, y

        if (event.cell[1] >= this.cells.y + this.cells.rows)
        {
            return
        }
        if (this.dragStart)
        {
            steps = 1
            if (event.shift)
            {
                steps *= 2
            }
            if (event.ctrl)
            {
                steps *= 2
            }
            if (event.alt)
            {
                steps *= 2
            }
            var _a_ = this.state.mainCursor(); x = _a_[0]; y = _a_[1]

            switch (event.dir)
            {
                case 'up':
                    y -= steps
                    break
                case 'down':
                    y += steps
                    break
                case 'left':
                    x -= 1
                    break
                case 'right':
                    x += 1
                    break
            }

            y = _k_.clamp(0,this.state.s.lines.length - 1,y)
            x = _k_.clamp(0,this.state.s.lines[y].length - 1,x)
            start = [this.dragStart[0],this.dragStart[1]]
            if (y < this.dragStart[1])
            {
                start = [this.dragStart[2],this.dragStart[1]]
            }
            if (this.state.select(start,[x,y]))
            {
                this.redraw()
            }
            return
        }
        return fileeditor.__super__.onWheel.call(this,event)
    }

    return fileeditor
})()

export default fileeditor;

File: kakao.app/love/lua/view/editor/fileeditor.lua
--[[
00000000  000  000      00000000  00000000  0000000    000  000000000   0000000   00000000 
000       000  000      000       000       000   000  000     000     000   000  000   000
000000    000  000      0000000   0000000   000   000  000     000     000   000  0000000  
000       000  000      000       000       000   000  000     000     000   000  000   000
000       000  0000000  00000000  00000000  0000000    000     000      0000000   000   000
--]]

editor = require "edit.editor"
context = require "view.menu.context"


local fileeditor = class("fileeditor", editor)
    


function fileeditor:init(name) 
        local features = array('scroll', 'gutter', 'mapscr', 'complete', 'filepos', 'replex', 'brckts', 'unype', 'salter', 'vimple', 'uniko')
        
        editor.init(self, name, features)
        
        -- if @feats.mapscr
        --     @mapscr = mapscr @
        --     @mapscr.show()
        
        post:on('editor.highlight', self.state.highlightText, self.state)
        post:on('goto.line', self.onGotoLine, self)
        post:on('goto.bof', self.onGotoBof, self)
        post:on('goto.eof', self.onGotoEof, self)
        post:on('goto.func', self.onGotoFunc, self)
        post:on('git.diff', self.onGitDiff, self)
        post:on('git.commit', self.onGitCommit, self)
        post:on('funtree.loaded', self.onFuntreeLoaded, self)
        return self
    end


function fileeditor:setCurrentFile(currentFile) 
        self.currentFile = currentFile
        self.cells.meta_clear()
        return self.gutter.clearChanges()
    end


function fileeditor:onFuntreeLoaded() 
        if valid(self.gotoFuncOnLoad) then 
            self:onGotoFunc(self.gotoFuncOnLoad)
            return delete(self.gotoFuncOnLoad)
        end
    end


function fileeditor:onGitCommit() 
        return self.gutter.clearChanges()
    end

--  ███████   ███  █████████     ███████    ███  ████████  ████████
-- ███        ███     ███        ███   ███  ███  ███       ███     
-- ███  ████  ███     ███        ███   ███  ███  ██████    ██████  
-- ███   ███  ███     ███        ███   ███  ███  ███       ███     
--  ███████   ███     ███        ███████    ███  ███       ███     


function fileeditor:onGitDiff(diff) 
        local currentFile = ked_session.get('editor▸file')
        if (diff.file == currentFile) then 
            return self.gutter.onGitDiff(diff)
        end
    end

--[[        
     ███████    ███████   █████████   ███████         ███      ███  ███   ███  ████████  
    ███        ███   ███     ███     ███   ███        ███      ███  ████  ███  ███       
    ███  ████  ███   ███     ███     ███   ███        ███      ███  ███ █ ███  ███████   
    ███   ███  ███   ███     ███     ███   ███        ███      ███  ███  ████  ███       
     ███████    ███████      ███      ███████         ███████  ███  ███   ███  ████████  
    
    sets the main cursor:
        ◆ row       zero based line index to move the main cursor to
        ◆ col       optional column for the main cursor
        ◆ view      optional view offset
    --]]


function fileeditor:onGotoLine(row, col, view) 
        local mc = self.state.mainCursor()
        
        col = col or (mc[0])
        if is(col, str) then 
            if (col == 'ind') then col = belt.numIndent(self.state.s.lines[row])
            end
        end
        
        if (valid(view) and is(view, array)) then 
            self.state.setView(view)
            return self.state.setCursors(array(array(col, row)))
        else 
            local adjust = (view or 'topBotDeltaGrow')
            return self.state.setCursors(array(array(col, row)), {adjust = adjust})
        end
    end


function fileeditor:onGotoBof() 
    return self.state.moveCursors('bof')
    end

function fileeditor:onGotoEof() 
    return self.state.moveCursors('eof')
    end


function fileeditor:onGotoFunc(func) 
        local li = self.funtree.lineIndexForFunc(func)
        if li then 
            return self:onGotoLine(li, 'ind', 'topDelta')
        end
    end

--  ███████   ███████   ███   ███  █████████  ████████  ███   ███  █████████
-- ███       ███   ███  ████  ███     ███     ███        ███ ███      ███   
-- ███       ███   ███  ███ █ ███     ███     ███████     █████       ███   
-- ███       ███   ███  ███  ████     ███     ███        ███ ███      ███   
--  ███████   ███████   ███   ███     ███     ████████  ███   ███     ███   


function fileeditor:onContext(event) 
        local word = self.state:textOfSelectionOrWordAtCursor()
        if valid(word) then word = " '" .. tostring(word) .. "'" end
        return context.show(event.cell, fileeditor.onContextChoice, array("search" .. tostring(word) .. "", "find" .. tostring(word) .. "", 'status'))
    end


function fileeditor.static.onContextChoice(choice) 
        if kstr.startsWith(choice, 'search') then 
            return post:emit('searcher.show', kstr.trim(string.sub(choice, 7, -2), " '"))
        elseif kstr.startsWith(choice, 'find') then 
            return post:emit('finder.show', kstr.trim(string.sub(choice, 5, -2), " '"))
        else 
            if (choice == 'status') then 
    return post:emit('differ.status')
            end
        end
    end

--       ███  ███   ███  ██     ██  ████████         █████████   ███████         ███   ███   ███████   ████████   ███████  
--       ███  ███   ███  ███   ███  ███   ███           ███     ███   ███        ███ █ ███  ███   ███  ███   ███  ███   ███
--       ███  ███   ███  █████████  ████████            ███     ███   ███        █████████  ███   ███  ███████    ███   ███
-- ███   ███  ███   ███  ███ █ ███  ███                 ███     ███   ███        ███   ███  ███   ███  ███   ███  ███   ███
--  ███████    ███████   ███   ███  ███                 ███      ███████         ██     ██   ███████   ███   ███  ███████  


function fileeditor:jumpToWord(word) 
        -- log "fileeditor.jumpToWord '#{word}'"
        -- log 'fileeditor.jumpToWord classes' Object.keys(indexer.singleton.classes)
        -- log 'fileeditor.jumpToWord funcs'   Object.keys(indexer.singleton.funcs)
        
        local clss = indexer.singleton.classes[word]
        if clss then 
            post:emit('file.open', clss.file, (clss.line - 1))
            return true
        elseif indexer.singleton.funcs[word] then 
            local fnc = indexer.singleton.funcs[word]
            if is(fnc, array) then 
                local currentFile = ked_session.get('editor▸file')
                for fun, idx in ipairs(fnc) do 
                    if (fun.file == currentFile) then 
                        fnc = fnc[((idx + 1) % #fnc)]
                        break
                    end
                end
                
                if is(fnc, array) then 
                    fnc = fnc[0]
                end
            end
            
            if valid(fnc.file) then 
                post:emit('file.open', fnc.file, (fnc.line - 1), 'ind')
                return true
            end
        else 
            print("fileeditor.jumpToWord(" .. tostring(word) .. ") nothing found to jump to")
        end
        
        return false
    end


function fileeditor:singleCursorAtLine(li) 
        if empty(li) then return end
        
        self.state.setCursors(array(array(0, li)), {main = 'ind', adjust = 'topBotDelta'})
        
        return post:emit('redraw')
    end

--  ███████   ███████   ███   ███  ███   ███  █████████  ████████  ████████   ████████    ███████   ████████   █████████
-- ███       ███   ███  ███   ███  ████  ███     ███     ███       ███   ███  ███   ███  ███   ███  ███   ███     ███   
-- ███       ███   ███  ███   ███  ███ █ ███     ███     ███████   ███████    ████████   █████████  ███████       ███   
-- ███       ███   ███  ███   ███  ███  ████     ███     ███       ███   ███  ███        ███   ███  ███   ███     ███   
--  ███████   ███████    ███████   ███   ███     ███     ████████  ███   ███  ███        ███   ███  ███   ███     ███   


function fileeditor:jumpToCounterpart() 
        local currentFile = ked_session.get('editor▸file')
        local currext = slash.ext(currentFile)
        
        for ext in (fileutil.counterparts[currext] or array()) do 
            if nfs.fileExists(slash.swapExt(currentFile, ext)) then 
                self.gotoFuncOnLoad = self.funtree.current()
                post:emit('file.open', slash.swapExt(currentFile, ext))
                return
            end
        end
        
        for ext in (fileutil.counterparts[currext] or array()) do 
            local counter = slash.swapExt(currentFile, ext)
            local file = fileutil.swapLastDir(counter, currext, ext)
            
            if nfs.fileExists(file) then 
                self.gotoFuncOnLoad = self.funtree.current()
                post:emit('file.open', file)
                return
            end
        end
        
        for ext in (fileutil.counterparts[currext] or array()) do 
            local counter = slash.swapExt(currentFile, ext)
            
            if (currext == 'noon') then 
                local file = fileutil.swapLastDir(counter, 'kode', 'js')
                if nfs.fileExists(file) then 
                    post:emit('file.open', file)
                    return
                end
            end
            
            if (currext == 'json') then 
                local file = fileutil.swapLastDir(counter, 'js', 'kode')
                if nfs.fileExists(file) then 
                    post:emit('file.open', file)
                    return
                end
            end
        end
        
        return print('cant find counterpart', currentFile)
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function fileeditor:onMouse(event) 
        if editor.onMouse(self, event) then return true end
        if not ((self.dragStart or self.cells:isInsideEvent(event)) or self.gutter.cells:isInsideEvent(event)) then 
            return self.hover
        end
        
        local col, row = unpack(self:eventPos(event))
        
        if (event.type == 'press') then 
                if (event.count > 1) then 
                    if (not event.shift and (event.button == 'left')) then self.state:deselect() end
                    
                    local x = (col + self.state.s.view[1])
                    local y = (row + self.state.s.view[2])
                    
                    self.state:clearHighlights()
                    
                    if (event.count == 2) then 
                        if event.alt then 
                            self.state:selectChunk(x, y)
                        else 
                            self.state:selectWord(x, y)
                        end
                    else 
                        self.state:selectLine(y)
                    end
                    
                    self.state:highlightSelection()
                    
                    self.dragStart = copy(self.state.s.selections[1])
                    
                    return true
                else 
                    local x = ((col + self.state.s.view[1]) - 1)
                    local y = ((row + self.state.s.view[2]) - 1)
                    
                    if (event.cmd or event.ctrl) then 
                        local word = belt.wordAtPos(self.state.s.lines, array(x, y))
                        if word then 
                            -- log "jumpToWord #{word}" event
                            if self:jumpToWord(word) then 
                                return
                            end
                        end
                    end
                    
                    self.dragStart = array(x, y, x)
                    
                    if (not event.shift and (event.button == 'left')) then self.state:deselect() end
                    if not event.alt then self.state:clearCursors() end
                    
                    if event.alt then 
                        self.state:addCursor(x, y)
                    else 
                        if (event.shift and (#self.state.s.cursors == 1)) then 
                            self.state:setMainCursorAndSelect(x, y)
                        else 
                            self.state:setMainCursor(x, y)
                        end
                    end
                    
                    self:grabFocus()
                    
                    return true
                end
        elseif (event.type == 'drag') then 
                if self.dragStart then 
                    local x = (col + self.state.s.view[1])
                    local y = (row + self.state.s.view[2])
                    
                    local start = array(self.dragStart[0], self.dragStart[1])
                    
                    if (y < self.dragStart[1]) then 
                        start = array(self.dragStart[2], self.dragStart[1])
                    end
                    
                    if event.shift then self.state:addRangeToSelectionWithMainCursorAtEnd(belt.rangeFromStartToEnd(start, array(x, y)))
                    else self.state:select(start, array(x, y))
                    end
                    
                    return true
                end
        elseif (event.type == 'release') then 
                self.dragStart = nil
        elseif (event.type == 'move') then 
                if self.hover then 
                    if (not self:hasFocus() and empty(view.currentPopup)) then 
                        self:grabFocus()
                    end
                    
                    post:emit('pointer', 'text')
                elseif self.gutter.cells:isInsideEvent(event) then 
                    post:emit('pointer', 'vertical-text')
                end
        end
        
        return self.hover
    end

-- 000   000  000   000  00000000  00000000  000      
-- 000 0 000  000   000  000       000       000      
-- 000000000  000000000  0000000   0000000   000      
-- 000   000  000   000  000       000       000      
-- 00     00  000   000  00000000  00000000  0000000  


function fileeditor:onWheel(event) 
        if (event.cell[1] >= (self.cells.y + self.cells.rows)) then return end
        
        if self.dragStart then 
            local steps = 1 -- should be 4 if not using a mouse pad -> config
            if event.shift then steps = steps * 2 end
            if event.ctrl then steps = steps * 2 end
            if event.alt then steps = steps * 2 end
            
            local x, y = self.state.mainCursor()
            
            if (event.dir == 'up') then y = y - steps
            elseif (event.dir == 'down') then y = y + steps
            elseif (event.dir == 'left') then x = x - 1
            elseif (event.dir == 'right') then x = x + 1
            end
            
            y = clamp(0, (self.state.s.lines:len() - 1), y)
            x = clamp(0, (self.state.s.lines[y]:len() - 1), x)
            
            local start = array(self.dragStart[0], self.dragStart[1])
            
            if (y < self.dragStart[1]) then 
                start = array(self.dragStart[2], self.dragStart[1])
            end
            
            if self.state.select(start, array(x, y)) then 
                self:redraw()
            end
            
            return
        end
        
        -- super event
        return self
    end

return fileeditor